drivers/ethernet/eth_w5500.c

```diff
@@ -342,28 +342,52 @@ static int w5500_set_config(const struct device *dev,
 {
 	uint8_t mode;
 	uint8_t mr = W5500_S0_MR_MF;
+	struct w5500_runtime *ctx = dev->data;
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_MAC_ADDRESS:
+		memcpy(ctx->mac_addr,
+			config->mac_address.addr,
+			sizeof(ctx->mac_addr));
+		/* override vendor bytes */
+		ctx->mac_addr[0] = WIZNET_OUI_B0;
+		ctx->mac_addr[1] = WIZNET_OUI_B1;
+		ctx->mac_addr[2] = WIZNET_OUI_B2;
```

Comment:
Why does it "override vendor byte"? Why can a user not set his own OUI?

Reply:
@jfischer-no correct me if i'm wrong but I think this is also what is done for other Ethernet drivers in Zephyr?

Reply:
It seems to be a rumor. What driver does it for ETHERNET_CONFIG_TYPE_MAC_ADDRESS?

Reply:
If I remember correctly then I tried to set my own OUI bytes from my application too, but these values were ignored by the W5500 chip.

Reply:
@jfischer-no as an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_enc28j60.c#L820

Reply:
If W5500 hardware uses one MAC and the Zephyr stack knows another MAC in filtering layer then the networking doesn't work at all. I thought that it is better to fix it in driver level in the same way as the underlying HW is doing and saving the future developer using W5500 from trying to debug and understand why the network is not working.

I thought about another alternitive to return from this function with some error code if user provided OUI doesn't match with Wiznet OUI, but this means that Zephyr application has to be aware of underlying hardware and know about Wiznet OUI bytes which are not exposed from Zephyr headers.

Reply:
I prefer overwriting as proposed in the PR in this case.

Reply:
> @jfischer-no as an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_enc28j60.c#L820

No, it is not done during ETHERNET_CONFIG_TYPE_MAC_ADDRESS call.

> If W5500 hardware uses one MAC and the Zephyr stack knows another MAC in filtering layer then the networking doesn't work at all. I thought that it is better to fix it in driver level in the same way as the underlying HW is doing and saving the future developer using W5500 from trying to debug and understand why the network is not working.
> 
> I thought about another alternitive to return from this function with some error code if user provided OUI doesn't match with Wiznet OUI, but this means that Zephyr application has to be aware of underlying hardware and know about Wiznet OUI bytes which are not exposed from Zephyr headers.

If this controller can not change MAC address then this driver should not implement ETHERNET_CONFIG_TYPE_MAC_ADDRESS. If OUI is not modifiable, then the driver should not write to these registers/addresses. Overwriting OUI part multiple times in the driver is just confusing and opaque to maintainers and future contributors.

Reply:
W5500 supports user assigned MAC. This is how random MAC works and how I'm using to send fixed MAC with ETHERNET_CONFIG_TYPE_MAC_ADDRESS from my application in production. W5500 ignores the OUI part in the MAC and sending whole MAC was implemented in the driver before I did my changes. I haven't checked from W5500 documentation by myself if it supports changing partial MAC. Somewhat I doubt that it supports it.

I agree that overwriting the OUI part in multiple places is not the best solution. But the original code was already confusing for me so I decided not to change or touch these parts as I wasn't sure why they were implemented in that was as they are in source code. This might need a separate PR to refactor the existing code which I didn't want to do in my PR.

What I could offer to do in my PR to improve the comments with the reasons why some parts of the code have done by my as they are, so that future maintainers and contributors could take this into account when refactoring the code.

Reply:
> W5500 supports user assigned MAC. This is how random MAC works and how I'm using to send fixed MAC with ETHERNET_CONFIG_TYPE_MAC_ADDRESS from my application in production. W5500 ignores the OUI part in the MAC and sending whole MAC was implemented in the driver before I did my changes. I haven't checked from W5500 documentation by myself if it supports changing partial MAC. Somewhat I doubt that it supports it.
>

"W5500 supports user assigned MAC" contradicts "W5500 ignores the OUI part in the MAC". Discussion here is not about random MAC but about implementation of ETHERNET_CONFIG_TYPE_MAC_ADDRESS setting. 

> I agree that overwriting the OUI part in multiple places is not the best solution. But the original code was already confusing for me so I decided not to change or touch these parts as I wasn't sure why they were implemented in the way as they are in source code. This might need a separate PR to refactor the existing code which I didn't want to do in my PR.
> 
> What I could offer to do in my PR is to improve the comments with the reasons why some parts of the code have done by my as they are, so that future maintainers and contributors could take this into account when refactoring the code.

You have not checked the manual but try to reuse some already dirty parts, it does not sound like the right approach.

Reply:
I have full respect for the original contributors and maintainers of this driver and thankful for the whole Zephyr project. I believe that there are reasons why every part of the code has evolved historically as it is in use today. Given the time constraints I had back then and have today this was the safest approach to reuse existing code from this driver and other drivers within Zephyr project (ETHERNET_CONFIG_TYPE_MAC_ADDRESS implementation for example) to achieve the goals for my project and not to touch or modify the existing code paths that I couldn't understand fully without having to dig into W5500 documentation details. I shared my solution in the hope that someone more knowledgeable in Zephyr and W5500 internals could find it useful or use the parts of my code to reimplement the missing functionality properly if needed.

Reply:
Tested again and it looks like that this time custom OUI bytes were accepted with my W5500 chip. Removed the override OUI bytes from this PR.

Reply:
To clarify one statement I made yesterday, I checked the W5500 datasheet and I was wrong and W5500 should support partial MAC update as all settings are stored in one common register block.